README.md
# EJSChart

100% Powerful, Clean & Functional Javascript Charts

Whether at home, a medium sized or enterprise venture, EJSCharts will seamlessly help you represent your data more clearly and efficiently.

Compatible With:

* IE6+
* Firefox
* Safari
* Chrome

## Features

### Interactive

Features such as Hints, Mouse Tracking, Mouse Events, Key Tracking, and Events, Zooming, Scrolling, and Crosshairs raise interactivity and user experience in web charting to a new level.

Too much data and not enough screen real estate? Show it all. Let your end users zoom in on the pieces they're most interested in. Axis locking for single axis zoom, scrolling and automatic axis scaling are all included.

### Ajax Driven Data

EJSChart supports XML-formatted data and loads data on the fly. New series can be added and data updated in real time without page reloads.

### Stackable Series

Multiple chart series can be stacked and combined to fit many charting needs.

### Plugin Free

100% pure Javascript Charting solution. No more worries of incompatible plugin versions or confusing security warnings. EJSChart is pure JavaScript and requires no client installation.

## Setup

### Implementation Instructions

- Download the ZIP file for this EJSCharts repository.
- Create a new directory in your webroot to house the EJSCharts code.
- Unzip the contents of the EJSCharts ZIP file you downloaded into the directory you created.
- Include the EJSChart.min.js file in the HEAD portion of your HTML file using the following code:

`<script src="./path-to-your-directory/EJSChart.min.js"></script>`

### Creating Your First Chart

- Create a DIV elmeent in your HTML page where you want the chart to display and give it an id.

`<div id="myChart"></div>`

- Add a script block to create the chart.

```
<script>
var chart = new EJSC.Chart("myChart");
</script>
```

- Add series to the chart (see the [Examples][examples] area for examples of adding series).

## License

Please see the [LICENSE.txt][license] file for license information.

[examples]: http://www.ejschart.com/examples.php?selected=examplesPage
[license]: ./LICENSE.txt

CHANGES.txt
Version 2.3.1

1. New Features

    * N/A

2. Bug Fixes

    * Fixed IE canvas support check for IE10+

Version 2.3

1. New Features

	* Added pointStyle property to LineSeries and AreaSeries
	* Added even_shading and odd_shading properties to LinearAxis
	* Added onUserEndZoom event to Chart
	* Vertical Axis Captions (check cross-browser)
	* Added legend_position, legend_height, and legend_width to Chart
	* Added setAxis method to Chart
	* New Series: ErrorSeries
	* New Series: AlarmSeries
	* New Series: OverUnderSeries
	* New Series: StepSeries
	* New Series: DonutSeries
	* New Axis: DateAxis
	
2. Bug Fixes

	* Fixed issues with displaying charts in IE9


Version 2.2

1. New Features

	* Added a public method to return an array of series in the chart
	* Added support for context menu
	* Added onBeforeDrawSeries and onAfterDrawSeries events
	* Added support for BarSeries points with NULL for Y data
	* Added support for LineSeries and ScatterSeries points with NULL for Y data
	* Added support for individual grid line and tick styles via the OnNeedsTicks event

2. Bug Fixes

	* Fixed issues with all series reloading data
	* Fixed support for AnalogGaugeSeries to use x_axis_formatter to format it's ticks
	* Fixed issues with charts not displaying in IE when inside a TABLE
	* Fixed issues with the NumberFormatter not formatting text correctly
	* Fixed issues with the LineSeries and AreaSeries not drawing correctly when zoomed in
	* Fixed issues with multiple BarSeries not spacing themselves correctly when the base axis data is dates
	* Fixed zooming issues with LogarithmicAxis
	* Fixed zooming issues with LinearAxis
	* Fixed issues in IE with hiding a chart with a container with a border
	* Fixed issues with BarSeries ranges
	* Fixed issues with removing BarSeries and PieSeries
	* Fixed default hint string for PieSeries
	* Fixed issues with point selection with BarSeries and FloatingBarSeries (using distance)
	* Fixed issues with hint data for StackedBarSeries
	* Fixed issues with VML sometimes not showing up in IE8 in STRICT mode
	* Fixed issues with exporting SVG when no axes are being used
	* Fixed issues with hint pointer z-index
	* Fixed default hint texts for AreaSeries, ScatterSeries, PieSeries, and TrendSeries
	* Fixed issues with unselecting point before hiding a series
	* Fixed issue with stock series when open or close attributes were null


Version 2.1

1. New Features

	* Added treeLegendRoot property to StackedBarSeries
	* Added left and right as valid property values for PieSeries.position and AnalogGaugeSeries.position
	* Added defaultColors, useColorArray and onBarNeedsColor to StackedBarSeries
	* Added the ability for StackedBarSeries to share a single color array among all owned series
	* Added support for IE8 RC1 (and most previous betas) in all modes
	* Added chart.setShowLegend method
	* Performance increases in IE when drawing multiple series and/or series with large numbers of points
	* Added support for gaps in LineSeries (use a blank string as Y)
	* Added getBarSizeInPoints to BarSeries and descendant classes

2. Bug Fixes

	* Corrected several IE related memory leaks
	* Corrected memory leaks related to Ajax requests (affected all browsers)
	* Corrected memory leaks when removing charts from a page repeatedly without refreshing
	* Corrected parameter name for CSVStringDataHandler.setCSV
	* Fixed AreaSeries draw when series contains more than 5000 points in IE
	* Fixed incorrect parsing of numeric Y values in Stock Series
	* Updated chart.onBeforeSelectPoint, onAfterSelectPoint, onShowHint to send the correct value for the HoverOrSelect parameter
	* Corrected error when trying to get/set axis captions
	* Corrected issues with certain data handlers corrupting data when Object.prototype has been modified
	* Fixed issues with legend movement when the chart was inside a scrollable container
	* Corrected Axis.pointToPixel so that it returns the proper value when given a string (bin)
	* Corrected double loading of data when using JSONStringDataHandler
	* Removed the ability for closeLine to affect drawing in LineSeries
	* Corrected issues with AreaSeries drawing when closeLine was false in Safari
	* Corrected AreaSeries drawing in IE when more than 5000 points exist
	* Corrected issues with extremes calculations in StackedBarSeries


Version 2.0.1

1. New Features
	* Added Series.getPadding, Series.setPadding methods
	* Added Series.padding property
	* Added Chart.legend_state (normal, minimized)
	* Added Chart.legendMinimize, Chart.legendRestore, Chart.getLegendState methods
	* Added ignoreBounds parameter to Axis.pointToPixel

2. Bug Fixes
	* BarSeries / FloatingBarSeries - Fixed dimension calculation which caused bars
	  to be missed if the mouse was close to an axis
	* BarSeries / FloatingBarSeries - Updated point location / distance to return
	  exact matches when the mouse is within a bar
	* Fixed issue with chart files when in a directory named ejschart_
	* Corrected issue with axis.force_static_points and cursor position
	* Corrected issue with legend icons causing non-secure warning in https session
	* Added catch to all try/finally statements to account for IE bug
	* Corrected adjustment to so that crosshairs position exactly under the cursor
	* Renamed variable int to interval to correct Safari 2 parse issue
	* Updated default series padding to account for axis assignment
	* Corrected issue with horizontal bars not drawing when rendered 1 pixel wide

3. Documentation Updates
	* Added documentation for META tag options
	* Added documentation for Series.findClosestByPoint and Series.findClosestByPixel


Version 2.0

1. New Features:
	* Multiple X and Y Axes
	* Logarithmic Axis
	* JSON String and File Data Handlers
	* Stacked Bar Series
	* Floating Bar Series
	* String Formatter for prefixing and appending to labels
	* OpenHighLowClose Series
	* Candlestick Series
	* Filled Scatter Series Points in IE
	* Much optimized rendering of axis labels and ticks

help/HTML/ejschart_content.js
var tocTab = new Array();var ir=0;
tocTab[ir++] = new Array ("1", "Introduction", "", "", "cicon1.gif", "cicon2.gif");
tocTab[ir++] = new Array ("1.1", "Getting Started", "welcometopic.htm", "", "cicon9.gif", "cicon9.gif");
tocTab[ir++] = new Array ("1.2", "Data Formats", "dataformats.htm", "", "cicon9.gif", "cicon9.gif");
tocTab[ir++] = new Array ("1.3", "Customization", "customization.htm", "", "cicon9.gif", "cicon9.gif");
tocTab[ir++] = new Array ("1.4", "Examples &amp; Code Snippets", "examplescodesnippets.htm", "", "cicon9.gif", "cicon9.gif");
tocTab[ir++] = new Array ("1.5", "Additional Information", "additionalinformation.htm", "", "cicon9.gif", "cicon9.gif");
tocTab[ir++] = new Array ("1.6", "Changes", "", "", "cicon1.gif", "cicon2.gif");
tocTab[ir++] = new Array ("1.6.1", "Changes in version 0.6.0 Beta", "changesinversion0_6_0beta.htm", "", "cicon9.gif", "cicon9.gif");
tocTab[ir++] = new Array ("1.6.2", "Changes in version 0.5.3 Beta", "changesinversion0_5_3beta.htm", "", "cicon9.gif", "cicon9.gif");
tocTab[ir++] = new Array ("2", "API Reference", "", "", "cicon1.gif", "cicon2.gif");
tocTab[ir++] = new Array ("2.1", "EJSC", "ejscnamespace.htm", "", "cicon9.gif", "cicon9.gif");
tocTab[ir++] = new Array ("2.2", "EJSC.Chart", "ejsc_chart.htm", "", "cicon1.gif", "cicon2.gif");
tocTab[ir++] = new Array ("2.2.1", "Properties", "", "", "cicon1.gif", "cicon2.gif");
tocTab[ir++] = new Array ("2.2.1.1", "allow_interactivity", "ejsc_chart_properies_allow_int.htm", "", "cicon9.gif", "cicon9.gif");
tocTab[ir++] = new Array ("2.2.1.2", "allow_zoom", "ejsc_chart_properties_allow_zo.htm", "", "cicon9.gif", "cicon9.gif");
tocTab[ir++] = new Array ("2.2.1.3", "auto_find_point_by_x", "ejsc_chart_properties_auto_fin.htm", "", "cicon9.gif", "cicon9.gif");
tocTab[ir++] = new Array ("2.2.1.4", "force_static_points", "ejsc_chart_properties_force_st.htm", "", "cicon9.gif", "cicon9.gif");
tocTab[ir++] = new Array ("2.2.2", "Methods", "methods.htm", "", "cicon9.gif", "cicon9.gif");
tocTab[ir++] = new Array ("2.2.3", "Events", "ejsc_chart_events.htm", "", "cicon9.gif", "cicon9.gif");
tocTab[ir++] = new Array ("2.3", "Chart Series", "", "", "cicon1.gif", "cicon2.gif");
tocTab[ir++] = new Array ("2.3.1", "EJSC.LineSeries", "ejsc_lineseries.htm", "", "cicon9.gif", "cicon9.gif");
tocTab[ir++] = new Array ("2.3.2", "EJSC.AreaSeries", "ejsc_areaseries.htm", "", "cicon9.gif", "cicon9.gif");
tocTab[ir++] = new Array ("2.3.3", "EJSC.ScatterSeries", "ejsc_scatterseries.htm", "", "cicon9.gif", "cicon9.gif");
tocTab[ir++] = new Array ("2.3.4", "EJSC.PieSeries", "ejsc_pieseries.htm", "", "cicon9.gif", "cicon9.gif");
tocTab[ir++] = new Array ("2.3.5", "EJSC.BarSeries", "ejsc_barseries.htm", "", "cicon9.gif", "cicon9.gif");
tocTab[ir++] = new Array ("2.3.6", "EJSC.FunctionSeries", "ejsc_functionseries.htm", "", "cicon9.gif", "cicon9.gif");
tocTab[ir++] = new Array ("2.3.7", "EJSC.TrendSeries", "ejsc_trendseries.htm", "", "cicon9.gif", "cicon9.gif");
tocTab[ir++] = new Array ("2.4", "Data Handlers", "", "", "cicon1.gif", "cicon2.gif");
tocTab[ir++] = new Array ("2.4.1", "EJSC.XMLDataHandler", "ejsc_xmldatahandler.htm", "", "cicon9.gif", "cicon9.gif");
tocTab[ir++] = new Array ("2.4.2", "EJSC.ArrayDataHandler", "ejsc_arraydatahandler.htm", "", "cicon9.gif", "cicon9.gif");
tocTab[ir++] = new Array ("2.4.3", "EJSC.CSVFileDataHandler", "ejsc_csvfiledatahandler.htm", "", "cicon9.gif", "cicon9.gif");
tocTab[ir++] = new Array ("2.4.4", "EJSC.CSVStringDataHandler", "ejsc_csvstringdatahandler.htm", "", "cicon9.gif", "cicon9.gif");
tocTab[ir++] = new Array ("2.5", "Label Formatters", "", "", "cicon1.gif", "cicon2.gif");
tocTab[ir++] = new Array ("2.5.1", "EJSC.NumberFormatter", "ejsc_numberformatter.htm", "", "cicon9.gif", "cicon9.gif");
tocTab[ir++] = new Array ("2.5.2", "EJSC.DateFormatter", "ejsc_dateformatter.htm", "", "cicon9.gif", "cicon9.gif");
tocTab[ir++] = new Array ("2.6", "Interfaces", "", "", "cicon1.gif", "cicon2.gif");
tocTab[ir++] = new Array ("2.6.1", "EJSC.Inheritable", "ejsc_inheritable.htm", "", "cicon9.gif", "cicon9.gif");
tocTab[ir++] = new Array ("2.6.2", "EJSC.Series", "ejsc_series.htm", "", "cicon9.gif", "cicon9.gif");
tocTab[ir++] = new Array ("2.6.3", "EJSC.Point", "ejsc_point.htm", "", "cicon9.gif", "cicon9.gif");
tocTab[ir++] = new Array ("2.6.4", "EJSC.DataHandler", "ejsc_datahandler.htm", "", "cicon9.gif", "cicon9.gif");
tocTab[ir++] = new Array ("2.6.5", "EJSC.Formatter", "ejsc_formatter.htm", "", "cicon9.gif", "cicon9.gif");
isContent = true,
isIndex = false,
showNumbers = false,
tocBehaviour = new Array(1,1),
tocScroll=false,
tocLinks = new Array(0,0);
var oldOnError = null;
var isDyn = navigator.userAgent.toLowerCase().indexOf("opera") == -1;
var isIE = navigator.appName.toLowerCase().indexOf("explorer") > -1;
var oldCurrentNumber = "", oldLastVisitNumber = "";
var toDisplay = new Array();
for (ir=0; ir<tocTab.length; ir++) {
toDisplay[ir] = tocTab[ir][0].split(".").length==1;
}
function windowErr() {
isDyn = false;
hmtoc.location.href = "ejschart_content.htm";
window.onerror = oldOnError;
}
function reDisplay(currentNumber,tocChange,noLink,e) {
if (isIndex && (hmtoc.location.href.substring(hmtoc.location.href.lastIndexOf("/")+1,hmtoc.location.href.length) != "ejschart_kwindex.htm")) { isIndex=false; isContent=true; }
if (currentNumber == "navIndex") { isContent=false; }
if (currentNumber == "navContent") { isIndex=false; isContent=true; }
if (e) {
ctrlKeyDown = (isIE) ? e.ctrlKey : (e.modifiers==2);
if (tocChange && ctrlKeyDown) tocChange = 2;
}
var currentNumArray = currentNumber.split(".");
var currentLevel = currentNumArray.length-1;
var currentIndex = 1;
var currentHash = "";
if ((currentNumber == "") && (top.location.href.lastIndexOf("?") > 0)) {
currentNumber = top.location.href.substring(top.location.href.lastIndexOf("?")+1,top.location.href.length);
if (currentNumber.lastIndexOf(",") > 0) {
currentHash = "#" + currentNumber.substring(currentNumber.lastIndexOf(",")+1, currentNumber.length);
currentNumber = currentNumber.substring(0, currentNumber.lastIndexOf(","));
}
}
for (i=0; i<tocTab.length; i++) {
if ((tocTab[i][0] == currentNumber) || (tocTab[i][2] == currentNumber && tocTab[i][2] != "")) {
currentIndex = i;
currentNumber = tocTab[i][0];
currentNumArray = currentNumber.split(".");
currentLevel = currentNumArray.length-1;
break;
}
}
if (currentIndex < tocTab.length-1) {
nextLevel = tocTab[currentIndex+1][0].split(".").length-1;
currentIsExpanded = nextLevel > currentLevel && toDisplay[currentIndex+1];
}
else currentIsExpanded = false;
theHref = (noLink) ? "" : tocTab[currentIndex][2] + currentHash;
theTarget = tocTab[currentIndex][3];
for (i=1; i<tocTab.length; i++) {
if (tocChange) {
thisNumber = tocTab[i][0];
thisNumArray = thisNumber.split(".");
thisLevel = thisNumArray.length-1;
isOnPath = true;
if (thisLevel > 0) {
for (j=0; j<thisLevel; j++) {
isOnPath = (j>currentLevel) ? false : isOnPath && (thisNumArray[j] == currentNumArray[j]);
}
}
toDisplay[i] = (tocChange == 1) ? isOnPath : (isOnPath || toDisplay[i]);
if (thisNumber.indexOf(currentNumber+".")==0 && thisLevel > currentLevel) {
if (currentIsExpanded) toDisplay[i] = false;
else toDisplay[i] = (thisLevel == currentLevel+1);
}
}
}
if (!isContent && !isIndex) {
hmtoc.location.href = "ejschart_kwindex.htm";
isIndex = true;
}
hasSelection = false;
if (isContent) {
if (!isDyn) {
if (hmtoc.location.href.substring(hmtoc.location.href.lastIndexOf("/")+1,hmtoc.location.href.length) != "ejschart_content.htm") hmtoc.location.href = "ejschart_content.htm";
}
else {
oldOnError = window.onerror;
window.onerror = windowErr;
hmtoc.document.clear();
hmtoc.document.write("<html>\n\r<head></head>\n\r<style type=\"text/css\">\n\r       SPAN.heading1 { font-family: Arial,Helvetica; font-weight: normal; font-size: 10pt; color: #000000; text-decoration: none }\n\r       SPAN.heading2 { font-family: Arial,Helvetica; font-weight: normal; font-size: 9pt; color: #000000; text-decoration: none }\n\r       SPAN.heading3 { font-family: Arial,Helvetica; font-weight: normal; font-size: 8pt; color: #000000; text-decoration: none }\n\r       SPAN.heading4 { font-family: Arial,Helvetica; font-weight: normal; font-size: 8pt; color: #000000; text-decoration: none }\n\r       SPAN.heading5 { font-family: Arial,Helvetica; font-weight: normal; font-size: 8pt; color: #000000; text-decoration: none }\n\r       SPAN.heading6 { font-family: Arial,Helvetica; font-weight: normal; font-size: 8pt; color: #000000; text-decoration: none }\n\r\n\r       SPAN.hilight1 { font-family: Arial,Helvetica; font-weight: normal; font-size: 10pt; color: #FFFFFF; background: #002682; text-decoration: none }\n\r       SPAN.hilight2 { font-family: Arial,Helvetica; font-weight: normal; font-size: 9pt; color: #FFFFFF; background: #002682; text-decoration: none }\n\r       SPAN.hilight3 { font-family: Arial,Helvetica; font-weight: normal; font-size: 8pt; color: #FFFFFF; background: #002682; text-decoration: none }\n\r       SPAN.hilight4 { font-family: Arial,Helvetica; font-weight: normal; font-size: 8pt; color: #FFFFFF; background: #002682; text-decoration: none }\n\r       SPAN.hilight5 { font-family: Arial,Helvetica; font-weight: normal; font-size: 8pt; color: #FFFFFF; background: #002682; text-decoration: none }\n\r       SPAN.hilight6 { font-family: Arial,Helvetica; font-weight: normal; font-size: 8pt; color: #FFFFFF; background: #002682; text-decoration: none }\n\r</style>\n\r<body bgcolor=\"#FFFFFF\">\n\r<font face=\"Arial,Helvetica\" size=\"4\"><b>Title of this help project</b></font><br>\n\r<font face=\"Arial,Helvetica\" size=\"2\">\n\r<a href=\"javaScript:parent.reDisplay(\'navIndex\',0,0);\">Keyword Index</a>\n\r</font><br><br><br>\n\r\n\r  <!-- Place holder for the TOC, do not delete the line below -->\n\r  ");
for (i=0; i<tocTab.length; i++) {
if (toDisplay[i]) {
thisNumber = tocTab[i][0];
thisNumArray = thisNumber.split(".");
thisLevel = thisNumArray.length-1;
isCurrent = (i == currentIndex);
if (i < tocTab.length-1) {
nextLevel = tocTab[i+1][0].split(".").length-1;
img = (thisLevel >= nextLevel) ? tocTab[i][4] : ((toDisplay[i+1]) ? tocTab[i][5] : tocTab[i][4]);
}
else img = tocTab[i][4];
thisTextClass = ((thisNumber==currentNumber)?("hilight"):("heading"));
thisNode = ((thisNumber==currentNumber)?("<div id=currentNode></div>"):(""));
if (thisNumber==currentNumber) hasSelection = true;
hmtoc.document.write(thisNode);
hmtoc.document.write("<table border=0 cellspacing=0 cellpadding=2>");
if (img!="") {
hmtoc.document.write("<tr><td width=" + ((thisLevel+1) * 20) + " align=right valign=top>");
hmtoc.document.write("<a href=\"javaScript:history.go(0)\" onMouseDown=\"parent.reDisplay('" + thisNumber + "'," + tocBehaviour[0] + "," + tocLinks[0] + ",event)\">");
hmtoc.document.write("<img src=\"" + img + "\" border=0></a>");
}
else hmtoc.document.write("<tr><td width=" + (thisLevel * 20) + " align=right valign=top>");
hmtoc.document.write("</td><td align=left>");
hmtoc.document.write("<a href=\"javaScript:history.go(0)\" onMouseDown=\"parent.reDisplay('" + thisNumber + "'," + tocBehaviour[1] + "," + tocLinks[1] + ",event)\">");
hmtoc.document.write("<span class=\""  + thisTextClass + ((thisLevel>5) ? 6 : thisLevel+1) + "\">");
hmtoc.document.write( ((showNumbers)?(thisNumber+" "):"") + tocTab[i][1]);
hmtoc.document.write("</span></a>");
hmtoc.document.writeln("</td></tr></table>");
window.onerror = oldOnError;
} //isDyn
} //isContent
}
}
if (!noLink) {
oldLastVisitNumber = oldCurrentNumber;
oldCurrentNumber = currentNumber;
}
if (isContent && isDyn) {
hmtoc.document.write("\n\r\n\r<hr><font face=\"Arial,Helvetica\" size=\"1\">&copy; 2007 ... Your company</font>\n\r</body>\n\r</html>\n\r");
hmtoc.document.close();
}
if (theHref)
if (theTarget=="top") top.location.href = theHref;
else if (theTarget=="parent") parent.location.href = theHref;
else if (theTarget=="blank") open(theHref,"");
else {
hmcontent.location.href = theHref;
if ((tocScroll) && (hasSelection)) hmtoc.document.all["currentNode"].scrollIntoView();
}
}
